### PR TITLE
Fix bad duals in MPB interface

### LINKED ISSUE #36: KNITRO gives bad duals
Currently:
```
getreducedcosts(m::KnitroMathProgModel) = zeros(m.numVar)
getconstrduals(m::KnitroMathProgModel) = zeros(m.numConstr)
```
Can we return the actual lagrange multipliers instead? JuMP now has tests for nonlinear duals which KNITRO is failing.

### PR DESCRIPTION
**Pull request by yeesian**
closes #36

### CODE REVIEW

**Comment by yeesian** [comment]
Following the tests in https://github.com/JuliaOpt/JuMP.jl/blob/master/test/nonlinear.jl#L316-L394, here's what I ran:

```julia
# [nonlinear] Test nonlinear duals (Min)
using JuMP, KNITRO, FactCheck
modA = Model(solver=KnitroSolver())
@defVar(modA, x >= 0)
@defVar(modA, y <= 5)
@defVar(modA, 2 <= z <= 4)
@defVar(modA, 0 <= r[i=3:6] <= i)
@setNLObjective(modA, Min, -((x + y)/2.0 + 3.0)/3.0 - z - r[3])
@addConstraint(modA, cons1, x+y >= 2)
@addConstraint(modA, cons2, sum{r[i],i=3:5} <= (2 - x)/2.0)
@addNLConstraint(modA, cons3, 7.0*y <= z + r[6]/1.9)
@fact solve(modA) --> :Optimal
@fact getObjectiveValue(modA) --> roughly(-5.8446115, 1e-6)
@fact getValue(x)       --> roughly(0.9774436, 1e-6)
@fact getValue(y)       --> roughly(1.0225563, 1e-6)
@fact getValue(z)       --> roughly(4.0, 1e-6)
@fact getValue(r)[3]    --> roughly(0.5112781, 1e-6)
@fact getValue(r)[4]    --> roughly(0.0, 1e-6)
@fact getValue(r)[5]    --> roughly(0.0, 1e-6)
@fact getValue(r)[6]    --> roughly(6.0, 1e-6)

# Row duals
@fact MathProgBase.getconstrduals(modA.internalModel) --> roughly([0.333333, -1.0, -0.0714286], 1e-6)

# Reduced costs
@fact MathProgBase.getreducedcosts(modA.internalModel) -->
    roughly([0.0, 0.0, -1.07143, 0.0, 1.0, 1.0, -0.037594], 1e-6)
```

and

```julia
# [nonlinear] Test nonlinear duals (Max)
using JuMP, KNITRO, FactCheck
modA = Model(solver=KnitroSolver())
@defVar(modA, x >= 0)
@defVar(modA, y <= 5)
@defVar(modA, 2 <= z <= 4)
@defVar(modA, 0 <= r[i=3:6] <= i)
@setNLObjective(modA, Max, ((x + y)/2.0 + 3.0)/3.0 + z + r[3])
@addConstraint(modA, cons1, x+y >= 2)
@addConstraint(modA, cons2, sum{r[i],i=3:5} <= (2 - x)/2.0)
@addNLConstraint(modA, cons3, 7.0*y <= z + r[6]/1.9)
@fact solve(modA) --> :Optimal
@fact getObjectiveValue(modA) --> roughly(5.8446115, 1e-6)
@fact getValue(x)       --> roughly(0.9774436, 1e-6)
@fact getValue(y)       --> roughly(1.0225563, 1e-6)
@fact getValue(z)       --> roughly(4.0, 1e-6)
@fact getValue(r)[3]    --> roughly(0.5112781, 1e-6)
@fact getValue(r)[4]    --> roughly(0.0, 1e-6)
@fact getValue(r)[5]    --> roughly(0.0, 1e-6)
@fact getValue(r)[6]    --> roughly(6.0, 1e-6)

# Row duals
@fact MathProgBase.getconstrduals(modA.internalModel) --> roughly(-[0.333333, -1.0, -0.0714286], 1e-6)

# Reduced costs
@fact MathProgBase.getreducedcosts(modA.internalModel) -->
    roughly(-[0.0, 0.0, -1.07143, 0.0, 1.0, 1.0, -0.037594], 1e-6)
```

I couldn't figure out what went wrong in the mapping between JuMP and MPB though

**Comment by mlubin** [comment]
Nice! Passes JuMP tests now?

File: src/KnitroSolverInterface.jl [hunk 1]
```diff
@@ -226,8 +226,8 @@ end
 getobjval(m::KnitroMathProgModel) = m.inner.obj_val[1]
 getsolution(m::KnitroMathProgModel) = m.inner.x
 getconstrsolution(m::KnitroMathProgModel) = m.inner.g
-getreducedcosts(m::KnitroMathProgModel) = zeros(m.numVar)
-getconstrduals(m::KnitroMathProgModel) = zeros(m.numConstr)
+getreducedcosts(m::KnitroMathProgModel) = -1 .* m.inner.lambda[m.numConstr+1:end]
```
Comment (mlubin) [inline src/KnitroSolverInterface.jl]: Pretty strange that knitro stores these all in a single vector

Comment (yeesian) [inline src/KnitroSolverInterface.jl]: Ctrl+f `lambda` in the `KTR_solve` API under http://www.artelys.com/tools/knitro_doc/3_referenceManual/callableLibrary/API.html#solving

**Comment by mlubin** [comment]
> I couldn't figure out what went wrong in the mapping between JuMP and MPB though

What do you mean by this?

**Comment by yeesian** [comment]
I'm still triggering the error messages in `getDual()`, and haven't tried
tracing what's happening. I'm afk, and will paste the error messages here
later

**Comment by yeesian** [comment]
> What do you mean by this?

Following on from the code above:

```julia
julia> getDual(x)
ERROR: Variable bound duals (reduced costs) not available. Check that the model was properly solved and no integer variables are present.
 in getDual at /Users/yeesian/.julia/v0.3/JuMP/src/JuMP.jl:277
```
which turns up

```julia
# Dual value (reduced cost) getter
function getDual(v::Variable)
    if length(v.m.redCosts) < MathProgBase.numvar(v.m)
        error("Variable bound duals (reduced costs) not available. Check that the model was properly solved and no integer variables are present.")
    end
    return v.m.redCosts[v.col]
end
```
Inspecting the reduced costs, it seems like they're not being reflected in the JuMP model:
```julia
julia> x.m.redCosts
0-element Array{Float64,1}

julia> MathProgBase.numvar(x.m)
7
```

**Comment by mlubin** [comment]
What's the length of the vector returned by ``getreducedcosts``?

**Comment by yeesian** [comment]
```julia
julia> MathProgBase.getreducedcosts(x.m.internalModel)
7-element Array{Float64,1}:
 -6.39515e-9
  3.4028e-9
  1.07143
 -1.03279e-9
 -1.0
 -1.0
  0.037594```

**Comment by mlubin** [comment]
Check that the dual is getting assigned in ``solvenlp``

**Comment by yeesian** [comment]
Ah, I wasn't on JuMP master, I think

**Comment by mlubin** [comment]
That would do it

**Comment by yeesian** [comment]
A'ight, works now

**Pull request merged**